### ISSUE TITLE
Decoder AMP training loss blow up

### ISSUE THREAD

**Issue by CiaoHe** [lucidrains/DALLE2-pytorch#97]
Just pin my training issue:

when set `amp = True`, the decoder's loss(unet1 & unet2) blowed up to 1K+. [loss <0.01 when amp declined]
Just curious, does the half-precision training mode harm the DDPM training?

<img width="1057" alt="image" src="https://user-images.githubusercontent.com/39040787/168443277-bb3b6c85-5d5b-48a0-9d78-e53e6449dfee.png">

as a ref: https://wandb.ai/hcaoaf/dalle2-decoder/runs/3sb0pumh?workspace=user-hcaoaf

**Comment by CiaoHe** [comment]
Config 
```
Unets:
  Unet1:
    dim: 128
    image_embed_dim: 512
    text_embed_dim: 512
    cond_dim: 128
    channels: 3
    dim_mults: [1, 2, 3, 4] 
    cond_on_text_encodings: True
  
  Unet2:
    dim: 64
    image_embed_dim: 512
    text_embed_dim: 512
    cond_dim: 128
    channels: 3
    dim_mults: [1, 2, 3, 4]
    cond_on_text_encodings: False

Decoder:
  image_sizes: [64, 256]
  timesteps: 1000
  image_cond_drop_prob: 0.1
  text_cond_drop_prob: 0.5
  condition_on_text_encodings: True
  # loss
  loss_type: 'l2'
  learned_variance: True
  # diff
  beta_schedule: 'cosine'
  predict_x_start: False
  predict_x_start_for_latent_diffusion: False # not to latent now
  lowres_downsample_first: True
  # blur
  blur_sigma: 0.1
  blur_kernel_size: 3
  clip_denoised: True
  clip_x_start: True

**Comment by lucidrains** [comment]
yea, i have some ideas about how to fix this (it could be the cross attention conditioning blocks)

may get to this next week, if i do not end up accepting a job offer that came my way (my open sourcing may come to an end, as the company is not favorable on it)

**Comment by lucidrains** [comment]
@CiaoHe how does it look without AMP? i also built in an unconditional feature for the decoder, so one can train it without cross attention conditioning

**Comment by lucidrains** [comment]
@CiaoHe i added an extra stability measure for the cross attention block - will get around to training the decoder next week (if i don't end up accepting this job offer that came my way :laughing: )

**Comment by lucidrains** [comment]
@CiaoHe also realized the default learning rate i had (3e-4) is a bit too high compared to Ho's original ddpm (2e-5)

**Comment by lucidrains** [comment]
worse comes to worse, i can bring in the cosine sim attention from SwinV2 for the cross attention blocks, if we diagnose that to be the issue

**Comment by CiaoHe** [comment]
> @CiaoHe how does it look without AMP? i also built in an unconditional feature for the decoder, so one can train it without cross attention conditioning (im guessing it is probably the attention blocks blowing up, so i added some additional normalization that should help)

here's w/o AMP. 
<img width="1057" alt="image" src="https://user-images.githubusercontent.com/39040787/168453920-51da1316-3d5a-4c84-b2a2-60f22154e072.png">

**Comment by lucidrains** [comment]
@CiaoHe how high is your batch size? i just added built-in gradient accumulation, so you can try for higher batch sizes without OOM

**Comment by CiaoHe** [comment]
@lucidrains  wow cool built-in grad-accum support.
Now, everything works fine (`max-batch-size`:32, `total-batch-size`:265, `amp`:True)
<img width="1057" alt="image" src="https://user-images.githubusercontent.com/39040787/168457179-8d7824cc-fd9d-45ec-8c1c-fb47425d3240.png">

**Comment by lucidrains** [comment]
@CiaoHe hurray 🎊 🎉

**Comment by CiaoHe** [comment]
I missed some msgs you replied in the github box. It will be a massive loss if u end up open source work.(ಥ_ಥ) But anyway hope you get a excellent experience!

获取Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
